File: src/platform/posix/appl.c
/*
 * Copyright 2014-2016, Björn Ståhl
 * License: 3-Clause BSD, see COPYING file in arcan source repository.
 * Reference: http://arcan-fe.com
 */

#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include <libgen.h>
#include <string.h>
#include <math.h>
#include <assert.h>
#include <ctype.h>

#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <arcan_math.h>
#include <arcan_general.h>

static size_t appl_len = 0;
static char* g_appl_id = "#appl not initialized";
static char* appl_script = NULL;

/*
 * This is also planned as an entry point for implementing the
 * APF mapping, possibly with the chainloader setting envvars
 * for how the namespace is made, and if not (.apf extension in appl_id)
 * setup the context and functions required to do data lookups.
 */
bool arcan_verifyload_appl(const char* appl_id, const char** errc)
{
	size_t app_len;

	if (!appl_id || (app_len = strlen(appl_id)) == 0){
		*errc = "no application specified";
		return false;
	}

/*
 * NOTE: here is the place to check for .fap extension,
 * and if found, take a separate route for setting up.
 */
	char* base = strdup(appl_id);
	bool expand = true;

/*
 * absolute/relative path? don't define namespaces using SYS_*
 * will maintain STATE namespace set by paths. Absolute paths
 * should only be presented from a privileged state (command-line
 * or in-engine, not Lua->engine).
 */
	if (appl_id[0] == '/' || appl_id[0] == '\\' || appl_id[0] == '.'){
		char* work = strdup(base);
		char* dir = strdup( dirname( work ) );
		free(work);

		work = strdup(base);
		free(base);
		base = strdup( basename(work) );
		free(work);

		arcan_override_namespace(appl_id, RESOURCE_APPL);

		arcan_softoverride_namespace(appl_id, RESOURCE_APPL_TEMP);
		arcan_softoverride_namespace(dir, RESOURCE_SYS_APPLBASE);
		arcan_softoverride_namespace(dir, RESOURCE_SYS_APPLSTORE);

		free(dir);
		expand = false;
	}

/*
 * with path stripped, we can make sure to see that we have a proper name
 */
	app_len = strlen(base);
	for (size_t i = 0; i < app_len; i++){
		if (!isalnum(base[i]) && base[i] != '_'){
			*errc = "invalid character in appl_id (only a..Z _ 0..9 allowed)";
			return false;
		}
	}

	if (expand){
		char* dir = arcan_expand_resource(base, RESOURCE_SYS_APPLBASE);
		if (!dir){
			*errc = "missing application base";
			free(base);
			return false;
		}
		arcan_override_namespace(dir, RESOURCE_APPL);
		arcan_mem_free(dir);

		dir = arcan_expand_resource(base, RESOURCE_SYS_APPLSTORE);
		if (!dir){
			*errc = "missing application temporary store";
			free(base);
			return false;
		}
		arcan_override_namespace(dir, RESOURCE_APPL_TEMP);
		arcan_mem_free(dir);
	}

/* state- storage has specific rules so it is always checked for expansion,
 * if STATE is a subpath of shared, we have global application state (i.e.
 * multiple scripts can use the same state storage for virtual machines),
 * otherwise we expand */
	char* p_a = arcan_expand_resource("", RESOURCE_APPL_SHARED);
	char* p_b = arcan_expand_resource("", RESOURCE_SYS_APPLSTATE);
	if (!p_b)
		arcan_softoverride_namespace(p_a, RESOURCE_APPL_STATE);
	else if (strncmp(p_a, p_b, strlen(p_a)) == 0){
		arcan_softoverride_namespace(p_b, RESOURCE_APPL_STATE);
	}
	else {
		arcan_mem_free(p_b);
		p_b = arcan_expand_resource(base, RESOURCE_SYS_APPLSTATE);
		arcan_softoverride_namespace(p_b, RESOURCE_APPL_STATE);
	}
	arcan_mem_free(p_a);
	arcan_mem_free(p_b);

	char wbuf[ app_len + sizeof(".lua")];
	snprintf(wbuf, sizeof(wbuf), "%s.lua", base);

	char* script_path = arcan_expand_resource(wbuf, RESOURCE_APPL);
	if (!script_path || !arcan_isfile(script_path)){
		*errc = "missing script matching appl_id (see appname/appname.lua)";
		arcan_mem_free(script_path);
		return false;
	}

/*
 * Switch to appl- suppled fonts if a folder exists to avoid relying on res
 * namespace pollution. Specific settings can still pin the namespace to
 * prevent this action.
 */
	char* font_path = arcan_expand_resource("/fonts", RESOURCE_APPL);
	if (font_path){
		if (arcan_isdir(font_path))
			arcan_override_namespace(font_path, RESOURCE_SYS_FONT);
		free(font_path);
	}

	if (strcmp(g_appl_id, "#appl not initialized") != 0)
		arcan_mem_free(g_appl_id);

	g_appl_id = base;
	appl_len = app_len;
	appl_script = script_path;

	return true;
}

const char* arcan_appl_basesource(bool* file)
{
	*file = true;
	return appl_script;
}

const char* arcan_appl_id()
{
	return g_appl_id;
}

size_t arcan_appl_id_len()
{
	return appl_len;
}

File: src/platform/posix/appl.zig
// Pure Zig port of posix/appl.c — zero C helpers.
// Application loader: verify and load appl (Lua script) by ID.

const std = @import("std");
const arcan = @import("arcan");

const c = struct {
    extern fn isalnum(c_arg: c_int) c_int;
    extern fn strdup(s: [*c]const u8) [*c]u8;
    extern fn free(ptr: ?*anyopaque) void;
    extern fn strcmp(a: [*c]const u8, b: [*c]const u8) c_int;
    extern fn strncmp(a: [*c]const u8, b: [*c]const u8, n: usize) c_int;
    extern fn strlen(s: [*c]const u8) usize;
    extern fn dirname(path: [*c]u8) [*c]u8;
    extern fn basename(path: [*c]u8) [*c]u8;
};

var appl_len: usize = 0;
var g_appl_id: [*c]const u8 = "#appl not initialized";
var appl_script: [*c]const u8 = null;

export fn arcan_verifyload_appl(appl_id: [*c]const u8, errc: *[*c]const u8) bool {
    if (appl_id == null) {
        errc.* = "no application specified";
        return false;
    }
    var app_len = c.strlen(appl_id);
    if (app_len == 0) {
        errc.* = "no application specified";
        return false;
    }

    var base: [*c]u8 = c.strdup(appl_id);
    var expand = true;

    if (appl_id[0] == '/' or appl_id[0] == '\\' or appl_id[0] == '.') {
        var work = c.strdup(base);
        const dir = c.strdup(c.dirname(work));
        c.free(work);

        work = c.strdup(base);
        c.free(base);
        base = c.strdup(c.basename(work));
        c.free(work);

        arcan.arcan_override_namespace(appl_id, arcan.RESOURCE_APPL);
        arcan.arcan_softoverride_namespace(appl_id, arcan.RESOURCE_APPL_TEMP);
        arcan.arcan_softoverride_namespace(dir, arcan.RESOURCE_SYS_APPLBASE);
        arcan.arcan_softoverride_namespace(dir, arcan.RESOURCE_SYS_APPLSTORE);

        c.free(dir);
        expand = false;
    }

    app_len = c.strlen(base);
    for (0..app_len) |i| {
        const ch: c_int = @intCast(base[i]);
        if (c.isalnum(ch) == 0 and base[i] != '_') {
            errc.* = "invalid character in appl_id (only a..Z _ 0..9 allowed)";
            return false;
        }
    }

    if (expand) {
        var dir = arcan.arcan_expand_resource(base, arcan.RESOURCE_SYS_APPLBASE);
        if (dir == null) {
            errc.* = "missing application base";
            c.free(base);
            return false;
        }
        arcan.arcan_override_namespace(dir, arcan.RESOURCE_APPL);
        arcan.arcan_mem_free(dir);

        dir = arcan.arcan_expand_resource(base, arcan.RESOURCE_SYS_APPLSTORE);
        if (dir == null) {
            errc.* = "missing application temporary store";
            c.free(base);
            return false;
        }
        arcan.arcan_override_namespace(dir, arcan.RESOURCE_APPL_TEMP);
        arcan.arcan_mem_free(dir);
    }

    const p_a = arcan.arcan_expand_resource("", arcan.RESOURCE_APPL_SHARED);
    var p_b = arcan.arcan_expand_resource("", arcan.RESOURCE_SYS_APPLSTATE);
    if (p_b == null) {
        arcan.arcan_softoverride_namespace(p_a, arcan.RESOURCE_APPL_STATE);
    } else if (p_a != null and c.strncmp(p_a, p_b, c.strlen(p_a)) == 0) {
        arcan.arcan_softoverride_namespace(p_b, arcan.RESOURCE_APPL_STATE);
    } else {
        arcan.arcan_mem_free(p_b);
        p_b = arcan.arcan_expand_resource(base, arcan.RESOURCE_SYS_APPLSTATE);
        arcan.arcan_softoverride_namespace(p_b, arcan.RESOURCE_APPL_STATE);
    }
    arcan.arcan_mem_free(p_a);
    arcan.arcan_mem_free(p_b);

    // Build "basename.lua" using bufPrintZ
    var wbuf: [256]u8 = undefined;
    const lua_name = std.fmt.bufPrintZ(&wbuf, "{s}.lua", .{
        @as([*c]const u8, base)[0..app_len],
    }) catch return false;
    _ = lua_name;

    const script_path = arcan.arcan_expand_resource(&wbuf, arcan.RESOURCE_APPL);
    if (script_path == null or !arcan.arcan_isfile(script_path)) {
        errc.* = "missing script matching appl_id (see appname/appname.lua)";
        arcan.arcan_mem_free(script_path);
        return false;
    }

    // Check for fonts directory
    var fbuf: [256]u8 = undefined;
    const font_name = std.fmt.bufPrintZ(&fbuf, "/fonts", .{}) catch null;
    _ = font_name;
    const font_path = arcan.arcan_expand_resource(&fbuf, arcan.RESOURCE_APPL);
    if (font_path != null) {
        if (arcan.arcan_isdir(font_path))
            arcan.arcan_override_namespace(font_path, arcan.RESOURCE_SYS_FONT);
        c.free(font_path);
    }

    if (c.strcmp(g_appl_id, "#appl not initialized") != 0)
        arcan.arcan_mem_free(@constCast(g_appl_id));

    g_appl_id = base;
    appl_len = app_len;
    appl_script = script_path;

    return true;
}

export fn arcan_appl_basesource(file: *bool) [*c]const u8 {
    file.* = true;
    return appl_script;
}

export fn arcan_appl_id() [*c]const u8 {
    return g_appl_id;
}

export fn arcan_appl_id_len() usize {
    return appl_len;
}
